hyperspy/io_plugins/usid_hdf5.py

```diff
@@ -0,0 +1,447 @@
+        If None - all Main Datasets will be read
+    verbose : bool, Optional. Default = False
+        Whether or not to print debugging statements
+    ignore_non_linear_dims : bool, Optional
+        If True, parameters that were varied non-linearly in the desired dataset will result in Exceptions.
+        Else, all such non-linearly varied parameters will be treated as linearly varied parameters and
+        a Signal object will be generated.
+
+    Returns
+    -------
+    list of hyperspy.signals.Signal object
+    """
+    if not isinstance(filename, str):
+        raise TypeError('filename should be a string')
+    if not os.path.isfile(filename):
+        raise FileNotFoundError('No file found at: {}'.format(filename))
+
+    with h5py.File(filename, mode='r') as h5_f:
+        if dset_path is None:
+            all_main_dsets = usid.hdf_utils.get_all_main(h5_f)
+            signals = []
+            for h5_dset in all_main_dsets:
+                # Note that the function returns a list already. Should not append
+                signals += __usidataset_to_signal(h5_dset, verbose=verbose,
+                                                  ignore_non_linear_dims=ignore_non_linear_dims)
+            return signals
+        else:
+            if not isinstance(dset_path, str):
+                raise TypeError('dset_path should be a string')
+            if len(dset_path) > 0:
+                h5_dset = h5_f[dset_path]
+                # All other checks will be handled by helper function
+            else:
+                all_main_dsets = usid.hdf_utils.get_all_main(h5_f)
+                if len(all_main_dsets) > 1:
+                    warn('{} contains multiple USID Main datasets.\n{}\nhas been selected as the desired dataset.'
+                         'If this is not the desired dataset, please supply the path to the main dataset via'
+                         ' the "dset_path" keyword argument.\nTo read all datasets, '
+                         'use "dset_path=None"'.format(h5_f, all_main_dsets[0].name))
+                h5_dset = all_main_dsets[0]
+            return __usidataset_to_signal(h5_dset, verbose=verbose, ignore_non_linear_dims=ignore_non_linear_dims)
+
+
+def file_writer(filename, object2save):
+    """
+    Writes a HyperSpy Signal object to a HDF5 file formatted according to USID
+
+    Parameters
+    ----------
+    filename : str
+        Path to target HDF5 file
+    object2save : hyperspy.signals.Signal
+        A HyperSpy signal
+    """
+    if not isinstance(filename, str):
+        raise TypeError('filename should be a string')
+    if os.path.exists(filename):
+        raise FileExistsError('Appending is not yet supported. A file already exists at: {}. '
+                              'Please delete the file at the location or specify a '
+                              'different path for the file'.format(filename))
+    """
+    # CANNOT import BaseSignal! Probably circular import
```

Comment:
Add it here, in this function and it should work fine.
```python
from hyperspy.signal import BaseSignal
```

Reply:
Turns out that I don't need to import the BaseSignal class after all so this is fine for now I guess. One question I have is how do I return a dataset (as a dictionary) which has N navigation axes and M signal axes? At the moment, this information is lost in the translation. When I [originally developed an ingestor](https://gist.github.com/ssomnath/097d2bb7c1468709ce2899512deb5b6d) - see lines 97 - 124, I was able to move the signal axes... I am not sure how to express this in the dictionary that I return now